src/libraries/System.Private.CoreLib/src/System/IO/RandomAccess.Unix.cs

```diff
@@ -173,33 +173,25 @@ internal static unsafe void WriteGatherAtOffset(SafeFileHandle handle, IReadOnly
 
             try
             {
-                int buffersOffset = 0, firstBufferOffset = 0;
-                while (true)
+                long totalBytesToWrite = 0;
+                for (int i = 0; i < buffersCount; i++)
                 {
-                    long totalBytesToWrite = 0;
-
-                    for (int i = buffersOffset; i < buffersCount; i++)
-                    {
-                        ReadOnlyMemory<byte> buffer = buffers[i];
-                        totalBytesToWrite += buffer.Length;
-
-                        MemoryHandle memoryHandle = buffer.Pin();
-                        vectors[i] = new Interop.Sys.IOVector { Base = firstBufferOffset + (byte*)memoryHandle.Pointer, Count = (UIntPtr)(buffer.Length - firstBufferOffset) };
-                        handles[i] = memoryHandle;
-
-                        firstBufferOffset = 0;
-                    }
+                    ReadOnlyMemory<byte> buffer = buffers[i];
+                    totalBytesToWrite += buffer.Length;
 
-                    if (totalBytesToWrite == 0)
-                    {
-                        break;
-                    }
+                    MemoryHandle memoryHandle = buffer.Pin();
+                    vectors[i] = new Interop.Sys.IOVector { Base = (byte*)memoryHandle.Pointer, Count = (UIntPtr)buffer.Length };
+                    handles[i] = memoryHandle;
+                }
 
+                int buffersOffset = 0;
+                while (totalBytesToWrite > 0)
+                {
                     long bytesWritten;
-                    Span<Interop.Sys.IOVector> left = vectors.Slice(buffersOffset);
+                    Span<Interop.Sys.IOVector> left = vectors.Slice(buffersOffset, buffersCount - buffersOffset);
```

Comment:
Would it make sense to change L171 to `stackalloc Interop.Sys.IOVector[IovStackThreshold].Slice(0, buffersCount)`, and then `vectors` will have Length == buffersCount and you won't need the second argument here, since vectors will always be correctly sized?

Reply:
It looks better, thanks!